contracts/ExponentialStaking.sol

```diff
@@ -0,0 +1,269 @@
+    /// else.
+    ///
+    /// If staking before the start of staking (epoch), then the lockup start
+    /// and end dates are shifted forward so that the lockup starts at the
+    /// epoch.
+    ///
+    /// Any rewards previously earned will be paid out or rolled into the stake.
+    ///
+    /// @param amountIn asset to lockup in the stake
+    /// @param duration in seconds for the stake
+    /// @param to address to receive ownership of the stake
+    /// @param stakeRewards should pending user rewards be added to the stake
+    /// @param lockupId previous stake to extend / add funds to. -1 to create a new stake.
+    function stake(uint256 amountIn, uint256 duration, address to, bool stakeRewards, int256 lockupId) external {
+        require(to != address(0), "Staking: To the zero address");
+        require(duration >= minStakeDuration, "Staking: Too short");
+        // Too long checked in preview points
+
+        uint256 newAmount = amountIn;
+        uint256 oldPoints = 0;
+        uint256 oldEnd = 0;
+        Lockup memory lockup;
+
+        // Allow gifts, but not control of other's accounts
+        if (to != msg.sender) {
+            require(stakeRewards == false, "Staking: Self only");
+            require(lockupId == NEW_STAKE, "Staking: Self only");
+        }
+
+        // Collect funds from user
+        if (amountIn > 0) {
+            // Important that `msg.sender` aways pays, not the `to` address.
+            asset.transferFrom(msg.sender, address(this), amountIn);
+            // amountIn already added into newAmount during initialization
+        }
+
+        // Collect funds from old stake (optional)
+        if (lockupId != NEW_STAKE) {
+            lockup = lockups[to][uint256(lockupId)];
+            uint256 oldAmount = lockup.amount;
+            oldEnd = lockup.end;
+            oldPoints = lockup.points;
+            require(oldAmount > 1, "Staking: Already closed stake");
+            emit Unstake(to, uint256(lockupId), oldAmount, oldEnd, oldPoints);
+            newAmount += oldAmount;
+        }
+
+        // Collect funds from rewards (optional)
+        newAmount += _collectRewards(to, stakeRewards);
+
+        // Caculate Points and lockup
+        require(newAmount > 0, "Staking: Not enough");
+        require(newAmount <= type(uint128).max, "Staking: Too much");
+        (uint256 newPoints, uint256 newEnd) = previewPoints(newAmount, duration);
+        require(newPoints + totalSupply() <= type(uint192).max, "Staking: Max points exceeded");
+        lockup.end = uint128(newEnd);
+        lockup.amount = uint128(newAmount); // max checked in require above
+        lockup.points = newPoints;
+
+        // Update or create lockup
+        if (lockupId != NEW_STAKE) {
+            require(newEnd > oldEnd, "Staking: New lockup must be longer");
+            lockups[to][uint256(lockupId)] = lockup;
+        } else {
+            lockups[to].push(lockup);
+            uint256 numLockups = lockups[to].length;
```

Comment:
The stake event doesn't emit the `lockupId` in case of a new stake correctly. Probably change this line to `lockupId = lockups[to].length`, that should work